### PR TITLE
Lora SFT example with DDP

### PR DESCRIPTION
**Pull request by philei-tt**
### Ticket
N/A

### Problem description
We need lora example with sft trainer + support for ddp in sft trainer

### What's changed
Small updates to sft trainer: 
- `on_before_optimizer_step` callback for gradient synchronization
- `loss_composer` parameter for multi-device loss aggregation in logging
- `attention_mask` parameter to pass an arbitrary mask (or None to let the model generate causal mask on the fly, which is faster)

Add train_lora_llama_sft.py -- reimplements the existing LoRA fine-tuning example using SFTTrainer + LoraConfig (HuggingFace PEFT-style API), with optional DDP and --loss_log for dumping per-step losses to JSON

Align data sampling in train_lora_llama.py -- replace random get_batch with shuffled-sequential chunk iteration matching InMemoryDataloader(shuffle=True), so both scripts produce identical losses

Update SFT_TRAINER.md with new constructor params, callback hook, and a DDP usage guide

I have verified that both sft and original script produce identical loss. 

### Checklist

- [ ] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:philei/sft_lora)

### CODE REVIEW

**Review by ayerofieiev-tt — APPROVED** [review]

File: tt-train/sources/ttml/ttml/trainers/sft_trainer.py [hunk 1]
```diff
@@ -182,13 +199,12 @@ def __init__(
...
-        self._lr_schedule = (
-            lr_schedule if lr_schedule is not None else self._build_lr_schedule()
-        )
-        self._causal_mask = self._build_causal_mask()
+        self._lr_schedule = lr_schedule if lr_schedule is not None else self._build_lr_schedule()
+        self._attention_mask = attention_mask
         self._loss_fn = ttml.ops.loss.cross_entropy_loss
         self._callbacks = callbacks or []
         self._compute_loss_override = compute_loss_func
+        self._loss_composer = loss_composer
```
Comment (dyurshevichTT) [inline tt-train/sources/ttml/ttml/trainers/sft_trainer.py]: I think we can always use the identity composer (even in single device case) so a user doesn't have to pass it here.

**Review by dyurshevichTT — APPROVED** [review]
good job!

**Review by jmalone-tt — APPROVED** [review]
Changes look good. I'm surprised it still takes so many lines to add train_lora_llama_sft.py - do you think we could cut this down by having a library of common dataset handlers, collate fn's, etc? Or is there something else that could help?
Thinking about the case of a customer adding a new training loop - ideally it is pretty concise

File: tt-train/docs/SFT_TRAINER.md [hunk 2]
```diff
@@ -79,6 +79,7 @@ SFTTrainer(
     lr_schedule=None,
     callbacks=None,
     compute_loss_func=None,
+    loss_composer=None,
 )
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/docs/SFT_TRAINER.md:84]: Fixed

File: tt-train/sources/ttml/ttml/trainers/sft_trainer.py (hunk 1 above)
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/trainers/sft_trainer.py]: Updated default loss composer to concatenate with dim=0

File: tt-train/sources/ttml/ttml/trainers/sft_trainer.py [hunk 3]
```diff
@@ -133,6 +139,8 @@ def __init__(
         lr_schedule: Optional[Callable[[int], float]] = None,
         callbacks: Optional[list[TrainerCallback]] = None,
         compute_loss_func: Optional[Callable] = None,
+        loss_composer: Any = None,
+        attention_mask: Any = None,
     ) -> None:
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/trainers/sft_trainer.py:146]: Good catch, fixed
